cpp/src/arrow/compute/kernels/cast.cc

```diff
@@ -460,6 +460,29 @@ struct CastFunctor<Date32Type, Date64Type> {
   }
 };
 
+// ----------------------------------------------------------------------
+// List to List
+
+template <>
+struct CastFunctor<ListType, ListType> {
+  void operator()(FunctionContext* ctx, const CastOptions& options,
+                  const ArrayData& input, ArrayData* output) {
+    Datum datum_out;
+    ListArray input_array(input.Copy());
+
+    const ListType& type = static_cast<const ListType&>(*output->type);
+
+    FUNC_RETURN_NOT_OK(
+        Cast(ctx, Datum(input_array.values()), type.value_type(), options, &datum_out));
+
+    FUNC_RETURN_NOT_OK(
+        input.buffers[0]->Copy(0, input.buffers[0]->size(), &output->buffers[0]));
+    FUNC_RETURN_NOT_OK(
+        input.buffers[1]->Copy(0, input.buffers[1]->size(), &output->buffers[1]));
+    output->child_data.emplace_back(datum_out.array());
```

Comment:
There's a couple issues with this:

a) We should not perform any additional copying or allocation beyond what is required to perform the cast on the child, i.e. we should simply reuse the buffers from the input array without copying

b) It would be better to resolve the cast kernel for the child data when the kernel is resolved in `GetCastFunction`. Instead of defining a functor for list, instead create an alternative implementation of `UnaryKernel` that has the child cast kernel as a member variable. This is what I meant in https://issues.apache.org/jira/browse/ARROW-1487 by "If a cast rule for the child type is not implemented, the failure should occur during kernel selection rather than at kernel evaluation time."